File: doit-csv/src/main/javaa/hr/element/doit/csvjava/CSVFactory.java
package hr.element.doit.csvjava;


import java.io.InputStream;
import java.nio.charset.Charset;

public class CSVFactory {
  public final String delimiter;
  public final String newLine;
  public final String quotes;
  public final Charset encoding;

  public static CSVFactory factory() {
    return new CSVFactory(";", "\n", "\"", Charset.forName("UTF-8"));
  }

  private CSVFactory(final String delimiter, final String newLine,
      final String quotes, final Charset encoding) {
    this.delimiter = delimiter;
    this.newLine = newLine;
    this.quotes = quotes;
    this.encoding = encoding;
  }

  public CSVFactory setDelimiter(final String delimiter) {
    return new CSVFactory(delimiter, newLine, quotes, encoding);
  }

  public CSVFactory setDelimiter(final char delimiter) {
    return setDelimiter(String.valueOf(delimiter));
  }

  public CSVFactory setNewLine(final String newLine) {
    return new CSVFactory(delimiter, newLine, quotes, encoding);
  }

  public CSVFactory setNewLine(final char newLine) {
    return setNewLine(String.valueOf(newLine));
  }

  public CSVFactory setEncoding(final String encoding) {
    return setEncoding(Charset.forName(encoding));
  }

  public CSVFactory setEncoding(final Charset encoding) {
    return new CSVFactory(delimiter, newLine, quotes, encoding);
  }

  public CSVFactory setQuotes(final String quotes) {
    return new CSVFactory(delimiter, newLine, quotes, encoding);
  }

  public CSVFactory setQuotes(final char quotes) {
    return setQuotes(String.valueOf(quotes));
  }

  // ----------------------------------------------------
/*
  public CSVWriter getWriter(final OutputStream oS) {
    Writer w = new OutputStreamWriter(oS, encoding);
    return new CSVWriter(this, w);
  }
*/
  // -----------------------------------------------------
  public CSVReaderOld getReader(final InputStream iS) {
    return new CSVReaderOld(this, iS);
  }
}

File: doit-csv/src/main/scala/hr/element/doit/csv/CSVFactory.scala
package hr.element.doit.csv

import java.io._
import java.nio.charset.Charset

object CSVFactory {

  def apply(): CSVFactory = {
    new CSVFactory("\"", ";", "\n", Charset.forName("UTF-8"))
  }
}

class CSVFactory private(
    val quotes: String,
    val delimiter: String,
    val newLine: String,
    val encoding: Charset){

  //----------------------------------------------------------------

  def getReader(in: InputStream) =
    CSVReader(this, in);

  def getReaderWithHeaders(in: InputStream) =
    CSVReader.withHeaders(this, in)

  def getWriter(out: OutputStream) =
    new CSVWriter(this, out)

  //----------------------------------------------------------------

  def setDelimiter(delimiter: String) =
    new CSVFactory(quotes, delimiter, newLine,encoding)

  def setDelimiter(delimiter: Array[Char]) =
    new CSVFactory(quotes, new String(delimiter), newLine,encoding)

  //----------------------------------------------------------------

  def setQuote(quotes: String) =
    new CSVFactory(quotes, delimiter, newLine,encoding)

  def setQuote(quotes: Array[Char]): CSVFactory =
    setQuote(new String(quotes))

  def setQuote(quotes: Char): CSVFactory =
    setQuote(quotes.toString)

  //----------------------------------------------------------------

  def setNewLine(newLine: String): CSVFactory =
    new CSVFactory(quotes, delimiter, newLine,encoding)

  def setNewLine(newLine: Array[Char]): CSVFactory =
    setNewLine(new String(newLine))

  def setNewLine(newLine: Char): CSVFactory =
    setNewLine(newLine.toString)

  //----------------------------------------------------------------

  def setEncoding(encoding: Charset): CSVFactory =
    new CSVFactory(quotes, delimiter, newLine,encoding)

  def setEncoding(encoding: String): CSVFactory =
    setEncoding(Charset.forName(encoding))

}

